_SPIKE_.md

```diff
@@ -0,0 +1,543 @@
+      index: true,
+      // unique: true
+    }
+  }
+}
+```
+
+### Indexes at model level
+
+Problem 1: different databases provide different indexing features.
+
+- It seems that all databases support ASC/DESC option at column-level.
+- MongoDB adds extra column-level option `text` as an alternative to ASC/DESC.
+- PostgreSQL allows column expressions, e.g. `lower(email)`.
+
+Problem 2: do we want to use database column names or LB property names? These
+two names can be different, especially when using database-specific config to
+map property names to custom column names.
+
+**The proposal:**
+
+At high-level, keep the current syntax where indexes are defined via a key-value
+map stored in `settings.indexes` property, the key is the index name and the
+value is an index definition object.
+
+```js
+{
+  strict: false,
+  forceID: true,
+  indexes: {
+    uniqueEmail: {
+      // index definition
+    },
+    nameQueries: {
+      // index definition
+    }
+  }
+}
+```
+
+Individual indexes can be defined as follows:
+
+- Add a new field `properties` as a key-value map from property names to
+  indexing order:
+
+  ```js
+  // definition of an individual index
+  {
+    properties: {
+      email: 1, // ASC
+      createdAt: 'DESC', // alias for -1
+      bio: 'text', // database-specific value (MongoDB's "text")
+    }
+  }
+  ```
+
+  Important: property names are mapped to database column names when building
+  the index definition.
+
+- Keep supporting `keys` field as a key-value map from database column name to
+  indexing order, see the description of the actual status below. Entries from
+  `keys` should me merged with entries from `properties`, `keys` taking
+  precedence (replacing `properties` entries).
+
+- Keep supporting `unique` field (set it to `true` to let the index enforce
+  uniqueness).
+
+- Database-specific options will be stored under a key with the connector name:
+
+  ```js
+  {
+    properties: {
+      email: 'ASC',
+    },
+    mongodb: {
+      sparse: true,
+    },
+    mysql: {
+      kind: 'fulltext',
+      type: 'hash',
+    },
+    postgresql: {
+      type: 'hash',
+    }
+  }
+  ```
+
+- To support more complex indexes offered by e.g. PostgreSQL, we should support
+  `columns` field too. This field should contain an array of column-index
+  specification, for example:
+
+  ```js
+  ['email ASC', 'lower(name) DESC'];
+  ```
+
+  In the contrary with the current implementation, I am proposing that `columns`
+  should replace any column list created from `properties` and `keys`, and that
+  `columns` should be nested inside connector-specific options:
+
+  ```js
+  {
+    properties: {
+      // default setup used by most connectors
+      email: 'ASC',
+    },
+    postgresql: {
+      // in PostgreSQL, use a different index definition
+      columns: ['lower(email) ASC']
+    },
+  }
+  ```
+
+### Foreign keys at property level
+
+Introduce a new property metadata "references" (inspired by ANSI SQL):
```

Comment:
I like the idea of references, but isn't this duplicating data? How can this example be applied to the TodoList example? In the todo model, we define todoListId as `belongsTo` (aka references) `TodoList`

https://github.com/strongloop/loopback-next/blob/master/examples/todo-list/src/models/todo.model.ts#L33

Maybe there is a way to leverage or combine this syntax with relational decorators?

Reply:
> I like the idea of references, but isn't this duplicating data?  How can this example be applied to the TodoList example?

I have modified the TodoList example to show how the new syntax could be used, see the bottom half of this pull request.

I feel we are not duplicating data here, because `belongsTo` decorator can be used without referential integrity, just to navigate to related models in GraphQL style. 

Also as we have discussed in the past, we need to rethink how relation decorators are applied. At the moment, `@hasMany` is applied on the navigational property (`todoList.todos`) while `@belongsTo` is applied on the foreign key (`todo.todoListId`) - this is inconsistent & possibly confusing. The navigational property does not belong to the model base class because relational data is not part of model data, create/update operations are not able to update related data in a single call. So far, we are leaning to rework decorators to be applied on the model constructor, and let the foreign-key property be defined independently of the relation.

Having said that, I can imagine that for now, we can modify `@belongsTo` decorator to add `references` metadata to the property definition it creates. The difficult part is how to infer `keyTo` (the primary key of the target model) at the time the decorator is invoked.

https://github.com/strongloop/loopback-next/blob/cb308add60bbe938c8e85812676eb817e9e0efc9/packages/repository/src/relations/belongs-to/belongs-to.decorator.ts#L26-L41

```ts
    if (!('references' in propertyDefinition)) {
      propertyDefinition.references = {
         model: targetResolver,
         property: keyTo,
      };
    }

    const propMeta: PropertyDefinition = Object.assign(
      // etc. (no change)
    );
```